File: src/org/intellij/markdown/lexer/GeneratedLexer.java
package org.intellij.markdown.lexer;

import org.intellij.markdown.IElementType;

public interface GeneratedLexer {
    void reset(CharSequence buffer, int start, int end,int initialState);

    IElementType advance() throws java.io.IOException;

    int getTokenStart();

    int getTokenEnd();

}

File: src/commonMain/kotlin/org/intellij/markdown/lexer/GeneratedLexer.kt
package org.intellij.markdown.lexer

import org.intellij.markdown.IElementType

interface GeneratedLexer {

    val tokenStart: Int

    val tokenEnd: Int
    fun reset(buffer: CharSequence, start: Int, end: Int, initialState: Int)

    fun advance(): IElementType?

    val state: Int
}


fun <E> ArrayList<E>.push(e: E) {
    add(e)
}

fun <E> ArrayList<E>.pop(): E {
    val result = last()
    dropLast(1)
    return result;
}
